### DISCUSSION TITLE
Question about concurrency (parallelism vs. context switching)

### DISCUSSION THREAD

**Discussion by paulzakin** [procrastinate-org/procrastinate#818] [Q&A] [answered]
Hello!

When you enable concurrency, say `procrastinate worker --concurrency=4`, what happens exactly? If you have a computer with a CPU with four cores - are all four cores saturated (e.g. parallelism)? Or is this flag just opening an event loop on one of the cores for asynchronous processing and (artificially) governing how many it can keep on the thread at one time (e.g. context switching)?  From some experimentation, I'm reasonably sure it is the latter, but just want to check. 

And if it is the latter, does true parallelism require running multiple versions of the worker itself? For example, running `python -m procrastinate --app=server.worker.app worker` in two separate processes on that same four-core machine.

Thanks!

PS: Love the library and our team has been experimenting with it to build a better mental model of how it works. Would you be open to a pull request from us for some improvements on documentation? Think it might make things a little easier for users like us that are new to this library and task queues!

**Comment by ewjoachim** [comment, marked as answer]
I think you got it right.

Even with an event loop, I/O is still happening in parallel, so if your tasks are I/O-bound, they're parallel. If your tasks are CPU-bound, the mechanisms within Procrastinate aren't going to be very helpful but, as you said, launching multiple processes will get you to use all the CPU.

The reason behind this choice is that multiprocessing is hard, and is often the place where a program crosses boundaries that make it more complex to deploy and scale. I felt the mental model of procrastinate was much simpler keeping it as a (mostly) single-threaded, single-processed program.

Note that those points are mentioned [here](https://procrastinate.readthedocs.io/en/stable/howto/concurrency.html#execute-multiple-jobs-at-the-same-time), and [here](https://procrastinate.readthedocs.io/en/stable/discussions.html#discussion-async) but if it was still confusing to you, then by all means, feel free to do a PR, and I'll do my best to review & merge it.

We're conceptually in the same situation as uvicorn, which lets you deploy a Python program that runs in an event loop. They have [multiple modes of deployment](https://www.uvicorn.org/deployment/):
- a basic `--worker n` argument that doesn't do process management
- an integration in Gunicorn, where Gunicorn gets to manage subprocesses, and uvicorn only needs to manage a single process
- an invitation to use the process manager of your stack to manage the different processes (e.g. SystemD, or k8s, or ...)

I think the idea is that (especially if you're managing an event loop already), it makes sense not to also manage child processes. People are going to be using a process manager of their choice, and trying to do anything from the lib itself will only get in their way and bring unwanted complexity.

**Reply by paulzakin** [reply]
This makes total sense - I think the Procrastinate model is very elegant and I agree with the choices made within - good to know my conclusion was mostly spot on!

(_The documentation thing was actually not concurrency related - it is what happens to all the “todo” status jobs of task X in the database when task X has a lock or queuing lock added to it by changing the decorator of the task when it did not have one before). We didn’t see any reference to what would happen in the documentation and were experimenting)_

**Reply by ewjoachim** [reply]
> it is what happens to all the “todo” status jobs of task X in the database when task X has a lock or queuing lock added to it by changing the decorator of the task when it did not have one before

Hm, I'm not 100% sure but from the top of my head, I'd say, that `lock` and `queueing_lock` are attached to the task in the DB (there are [corresponding columns](https://github.com/procrastinate-org/procrastinate/blob/b7fb36d39cc1d3397ae0aaf1c1ff67e7105784e8/procrastinate/sql/schema.sql#L30-L31)) and that's what triggers the behaviour, so I'd bet it's the value of the lock/queuing lock at the time the tasks were created that decides if they're going to be locked or not.

---

In a more general sense and for discussion sake, it's _possible_ (though generally not broadly advisable) that different workers define tasks differently. In your scenario, for example, you can imagine that you deploy some workers with the lock decorator, and keep other workers without it. That would certainly make for some horrible cases and I'm not sure what we should do about that or even if we should do something about that.

One thing that I've seen done here and there is people defining in the worker a cron-task and a queue that are unique to each worker, in a effort to have something periodically running on _every_ worker. So the worker knows about its own personal queue and its own personal task and not about all the other personal queues and tasks of the other workers. it kinda works, but I hate the headaches it creates when you need to update your mental model and imagine that there no official list of tasks, each worker actually has its own list of what it thinks are all the tasks, and we hope very much that they all agree :D

**Reply by paulzakin** [reply]
I agree with you 100% on that not being very advisable :) This is more handling the use case that we started a task without a lock and decided to add a lock a couple weeks after deployment and we wanted to know would happen to the old tasks. 

Broadly speaking, I guess this is the equivalent of a migration. In production we start with task A. We decide to modify task A after two weeks. What happens in the database - etc.

In a similar vein, let’s say we had task B, deployed it for two weeks, and then removed it. As we learned, the job will still remain in the database and Procrastinate will (correctly) throw an error about the lack of existence. So we need to delete all the old references to the job once the code is gone.

Effectively, we are building a guide internally to working with Procrastinate in production. Do you think other people might find that helpful? Maybe it could go here - https://procrastinate.readthedocs.io/en/stable/howto/deployment.html - which looks like it is a stub. Just a thought!

**Reply by paulzakin** [reply]
(Also, unrelated question, do you plan on bumping the version soon? I noticed that main is pretty far ahead now of the deployed version - we are quite happy with the current to be clear - I’m just curious!)

**Reply by ewjoachim** [reply]
> Broadly speaking, I guess this is the equivalent of a migration. In production we start with task A. We decide to modify task A after two weeks. What happens in the database - etc.

There are be counter-examples, but as a rule of thumb for the sake of creating one's mental model, I'd say: Tasks (the things defined in the code) are used to defer Jobs (the things we store in the DB). Jobs have some data stored with them, and when possible, we try to read data on the jobs rather than on the jobs' tasks. The important part of the task queue, locking mechanisms etc are handled by the DB itself, and for this to work, it needs to have all the relevant information (and the DB doesn't really know about tasks, it only knows about jobs. That's at least how it's structured in my mind but I can't say for sure it applies 100% of the time.

> Do you think other people might find that helpful?

Definitely !

> which looks like it is a stub

Mainly because it is. Sadly, I never really got to put procrastinate in production myself :D So feedback is appreciated. I always thought that we should absolutely get some observability features before considering that procrastinate is production ready (or it means every user needs to craft their own observability on top of procrastinate). That's the main reason why I never did the 1.0.

> do you plan on bumping the version soon?

Oh, I thought the only thing waiting to be released was dev env dependency bumps, but there are actual features waiting ! My bad !

**Reply by ewjoachim** [reply]
BTW, have you seen this ? https://github.com/procrastinate-org/procrastinate/discussions/748 :)

**Reply by paulzakin** [reply]
No worries about the release lol 

And I have seen #748! 

So we have an interest for sure, as we are (successfully) using Procrastinate in production. We are in the same boat as @jerluc in #748. However, I think if we can help a bit, we would like to. I think the caveat is this work would be pretty targeted towards our needs: improving the production readiness of Procrastinate (concurrency, observability, stuff like #780), documentation, and some ergonomic improvements. However, I'd imagine those improvements would benefit everyone! 

If all that sounds OK to you, maybe we could start with a code tour and introduction and go from there? We'd definitely need some guidance and architectural direction to help out, but it seems you'd be willing to provide that :) 

In general, I feel this library plugs an important hole in the Python ecosystem for people who do not want to use Redis (we love PostgreSQL as well) and do not want all the complexity of Celery. We've really enjoyed using it!

**Reply by ewjoachim** [reply]
> Honsetly, I feel this library plugs an important whole in the Python ecosystem

Aww that's nice :) Thanks ! <3

>  I think we can start by contributing some documentation and going from there?

Sure ! On my side, I'll do my best to make sure not to leave your questions unanswered :)

> we could start with a code tour and introduction and go from there?

Definitely ! Let's coordinate by Discord (`@ewjoachim`) ? (or email if you'd rather. my first name at my last name dot fr)

**Reply by paulzakin** [reply]
Sure, I'll ping you on Discord for further coordination!

**Reply by paulzakin** [reply]
Ok, cannot seem you to find you on Discord - I never use it lol so my settings are probably bad - let me try email - just sent you one.

**Reply by ewjoachim** [reply]
https://discord.gg/Svmp8wrf :)

**Reply by paulzakin** [reply]
Ok joined!

**Reply by ewjoachim** [reply]
(btw, https://github.com/procrastinate-org/procrastinate/releases/tag/0.28.0)

**Reply by paulzakin** [reply]
Yup, I saw - thanks a bunch!
